File: src/computer.rs
use crate::printer;
use crate::shell;
use std::fs;
use std::io::Read;
use std::thread;
use text_io::scan;

fn boot_system(
    cid: u16,
    mem_size: u32,
    time_quantum: u16,
    print_time: u16,
    pm_ip: String,
    pm_port: u16,
    n_comms: u16,
    conn_qsize: u16,
    msg_qsize: u16,
) {
    if cid != printer::printer_cid {
        memory::mem_init(mem_size);
        scheduler::scheduler_init(time_quantum);
        print::print_init(pm_ip, pm_port);
    } else {
        printer::printer_manager_init(pm_ip, pm_port, n_comms, conn_qsize, msg_qsize);
        printer::printer_init(print_time, n_comms);
    }
}

fn parse_config_params(mut config_f: fs::File) -> (String, u16, u32, u16, u16, u16, u16, u16) {
    let config_fmt: &str = "PM_IP:{}\nPM_PORT:{}\nM:{}\nTQ:{}\nPT:{}\nNC:{}\nCQS:{}\nMQS:{}";
    let mut pm_ip: String;
    let mut pm_port: u16;
    let mut mem_size: u32;
    let mut time_quantum: u16;
    let mut print_time: u16;
    let mut n_comms: u16;
    let mut conn_qsize: u16;
    let mut msg_qsize: u16;

    let mut config_str: String = String::new();
    config_f.read_to_string(&mut config_str);

    scan!(config_str.bytes() => config_fmt, pm_ip, pm_port, mem_size, time_quantum, print_time, n_comms, conn_qsize, msg_qsize);

    (
        pm_ip,
        pm_port,
        mem_size,
        time_quantum,
        print_time,
        n_comms,
        conn_qsize,
        msg_qsize,
    )
}

pub fn run(cid: u16) {
    let config_fname: String = String::from("config.sys");
    let config_f: fs::File = fs::File::open(config_fname).expect("Can't open config file.");

    let (pm_ip, pm_port, mem_size, time_quantum, print_time, n_comms, conn_qsize, msg_qsize) =
        parse_config_params(config_f);

    boot_system(
        cid,
        mem_size,
        time_quantum,
        print_time,
        pm_ip,
        pm_port,
        n_comms,
        conn_qsize,
        msg_qsize,
    );

    // Run shell and scheduler on client computers only
    if cid != printer::printer_cid {
        thread::spawn(shell::shell_operation);
        scheduler::process_execute();
    } else {
        thread::spawn(shell::printer_shell_operation);
        printer::printer_manager_main();
    }

    println!("Computer Run");
}

File: sim-os-c/project3/computer.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include "computer.h"

extern unsigned int PRINTER_CID;
unsigned int cid = 0;

void boot_system(unsigned int mem_size, unsigned int tq, unsigned int pt, char *pm_ip,
                 unsigned int pm_port, unsigned int n_comms, unsigned int conn_qsize, unsigned int msg_qsize)
{
    // Inititiate Memory and Scheduler on client computers only
    if (cid != PRINTER_CID)
    {
        mem_init(mem_size);
        process_init(tq);
        print_init(pm_ip, pm_port);
    }
    else
    {
        printer_manager_init(pm_ip, pm_port, n_comms, conn_qsize, msg_qsize);
        printer_init(pt, n_comms);
    }
}

int validate_cid(int read_cid)
{
    if (read_cid < 0)
    {
        fprintf(stderr, "Usage : ./computer.exe <CID>.\n");
        fprintf(stderr, "CID=%d revered for Printer, Other integer CIDs act as client computers.\n", PRINTER_CID);
        exit(1);
    }
    else if (read_cid == PRINTER_CID)
    {
        fprintf(stdout, "[computer.c] (main) : Computer (CID: %d) acting as Printer.\n", PRINTER_CID);
    }
    return read_cid;
}

int main(int argc, char **argv)
{
    if (argc == 2)
    {
        int read_cid = parse_int(argv[1]);
        cid = validate_cid(read_cid);
    }
    else
    {
        fprintf(stderr, "Usage : ./computer.exe <CID>\n");
        fprintf(stderr, "CID=%d revered for Printer, Other integer CIDs act as client computers.\n", PRINTER_CID);
        fprintf(stdout, "Input CID: ");
        scanf("%u", &cid);
        fprintf(stdout, "\n");

        cid = validate_cid(cid);
    }

    char *config_fname = "config.sys";
    FILE *config_fp = fopen(config_fname, "r");
    if (config_fp == NULL)
    {
        fprintf(stderr, "[computer.c] (main) : Can't open file %s\n", config_fname);
        return 1;
    }

    char pm_ip[16];
    unsigned int pm_port, mem_size, time_quantum, print_time, communicators, conn_qsize, msg_qsize;
    fscanf(config_fp, "PM_IP:%s\nPM_PORT:%d\nM:%d\nTQ:%d\nPT:%d\nNC:%d\nCQS:%d\nMQS:%d", pm_ip, &pm_port, &mem_size, &time_quantum, &print_time, &communicators, &conn_qsize, &msg_qsize);
    fclose(config_fp);

    boot_system(mem_size, time_quantum, print_time, pm_ip, pm_port, communicators, conn_qsize, msg_qsize);

    // Run shell and scheduler on client computers only
    if (cid != PRINTER_CID)
    {
        pthread_t shell_tid;
        pthread_create(&shell_tid, NULL, shell_operation, NULL);

        process_execute();
    }
    else
    {
        pthread_t printer_shell_tid;
        pthread_create(&printer_shell_tid, NULL, printer_shell_operation, NULL);

        printer_manager_main();
    }

    return 0;
}